Make EADL atomic relaxations consistent

mainegra opened the pull request:
This pull request addresses issues with the implementation of a more detailed atomic relaxation process and makes it officially the default in EGSnrc. Users wanting to revert back to the previous implementation will have to set the macro `$EADL_RELAX` to `.false.` in file `egsnrc.macros`. 
The mismatch between the set of binding energies used for photo-electric, Compton, and electron impact ionization (EII) interactions and the one used for the actual atomic relaxation creates a small inconsistency violating energy conservation. As reported by @FabioCpp in #277 , a simple simulation of a mono-energetic photon source in an infinite geometry reported deposited energy values that were different from the initial energy by almost 0.5%. Here at NRC a simulation of a 1 MeV photon source in an infinite geometry using `cavity` produced the following results for several elements and media:

|Medium  |  Z |   No EADL  |    EADL |
|-----------|-----|----------------|-------|
| **$EADL_RELAX**| | **.false.** | **.true.** |
|oxygen   |    8 |  1.00000000 |  0.9999992|
|manganese |  25 |  1.00000000 |  0.9999706|
|copper  |    29 |  1.00000000 |  0.9999541|
|tin        | 50  | 1.00000000 |  1.000112|
|terbium  |   65 |  0.99999921 |  _**1.000694**_|
|lead     |   82 |  0.99999831 |  _**1.001839**_|
|einsteinium |99 |  0.99992831 |  **_1.003197_**|
|water |       |    1.00000000  | 0.9999993|
|air     |       |  1.00000000 |  0.9999949|

Another issue with the current implementation is that the x-ray lines will correspond to the EADL radiative transitions regardless of the selection of the photon cross sections. Several changes and corrections have been made to resolve the energy conservation violation and make the x-ray lines correspond to the radiative transitions expected from the photon cross section compilation used. Here is a brief summary of the main changes:

- Make use of a consistent set of binding energies throughout the code. Binding energies are selected depending on the photon cross section libraries requested by the user, i.e., `xcom` or `epdl`. Transition probabilities are still taken from the EADL library. As a consequence, the position of the x-ray lines will match either the EADL radiative transitions (`epdl`) or the radiative transitions from the Table of Isotopes (8th Edition) (`xcom`).
- When `$EADL_RELAX` is true, allow all energetically possible transitions available in the EADL library.
- When `EADL_RELAX` is true, set `shn_array`, the shell type used during Compton interactions, to the local shell number rather than to the shell type as defined in the original implementation.
- When `$EADL_RELAX` is true, do not interact with \<M\> and \<N\> shells as done in  EGSnrc's default implementation for photo-absorption and relaxation. Characteristic x-ray lines generated from initital vacancies in the M shell have energies below 3 keV. See table below for more details on the shells considered by different interactions depending on the value of `$EADL_RELAX`:

|    Interaction   |          .false.           |               .true.                 |
|-------------------|-------------------------|---------------------------------|
| Compton         | K,L1..L3,M1..M5,N1..N7... |   K,L1..L3,M1..M5,N1..N7...   |
| EII                  | K,L1..L3                 | K,L1..L3                         |
| Photoeffect     | K,L1..L3,\<M\>,\<N\>  | K,L1..L3                         |
| Relaxation      |                               |                                      |
| initial vacancy | K,L1..L3,\<M\>          | K,L1..L3                        |
| final vacancy   |  L1..L3,\<M\>,\<N\>    | L1..L3,M1..M5,N1..N7... |

For more details on all the changes and corrections see the [commits](https://github.com/nrc-cnrc/EGSnrc/pull/282/commits) and [files changed](https://github.com/nrc-cnrc/EGSnrc/pull/282/files).

Here are the results for the total energy deposited per initial particle for different initial energies when using one of the two sets of binding energies consistently:

XCOM

|Medium |      Z |    0.1 MeV |    1 MeV |    10 MeV
|-----------|-------|---------------|------------|----------------|
|oxygen   |    8  | 0.10000000 | 1.0000000 | 10.000000 |
|manganese |  25 |  0.10000000 | 1.0000000 | 10.000000 |
|copper  |    29 |  0.10000000 | 1.0000000 | 10.000000 |
|tin        | 50 |  0.09999992 | 0.9999992 | 10.000000 |
|terbium   |  65 |  0.09999979 | 0.9999983 | 10.000000 |
|lead       | 82 |  0.09999967 | 1.0000000  | 9.999887|
|thorium    | 90 |  0.09999996 | 0.9999983  | 9.999900|
|einsteinium| 99 |  0.09999996 | 0.9999967 |  9.999983|
|water      |   |   0.10000000 | 1.0000000 | 10.000000|
|air           |  | 0.10000000 | 1.0000000 | 10.000000|

EPDL

|Medium  |     Z   |  0.1 MeV  |   1 MeV  |    10 MeV |
|------------|--------|--------------|------------|--------------|
|oxygen     |  8 |  0.10000000 | 1.0000000 | 10.000000 |
|manganese |  25 |  0.10000000 | 1.0000000 | 10.000000|
|copper  |    29 |  0.09999979 | 1.0000000 | 10.000000|
|tin        | 50 |  0.10000000 | 0.9999996 |  9.999995|
|terbium   |  65 |  0.09999967 | 0.9999992 |  9.999823|
|lead       | 82 |  0.09999983 | 0.9999983 |  9.999913|
|thorium    | 90 |  0.09999979 | 0.9999988 |  9.999963|
|einsteinium |99 |  0.09999979 | 0.9999992 |  9.999943|
|water       | |     0.10000000 | 1.0000000 | 10.000000|
|air           |   |0.10000000 | 1.0000000|  10.000000|

As can be seen there aren't values larger than the initial energy and the differences are negligible. This should fix issue #277.

The graph below shows W L-shell lines for a 20 kV x-ray beam obtained using the available options in EGSnrc: original ($EADL_RELAX = .false.), xcom or epdl photon cross sections ($EADL_RELAX = .true.). More lines are clearly seen for the detailed atomic relaxation implementation and a small shift in some lines between xcom and epdl is observed. This shift is likely irrelevant for energy deposition calculations and only of spectroscopical interest.
![20kv_al_spe_epdl_xcom_default_high](https://cloud.githubusercontent.com/assets/1971406/26425194/f3659000-40a1-11e7-9fe5-cb924d385c21.png)

ftessier commented:
Branch cleanup: rebase on current develop, reorder commits to pool small fixes at the end, squashed a couple commits, edited commit messages. No file content was changed, except formatting of the EADL_RELAX comment in `egsnrc.macros` and removing eol white space in `egsnrc.mortran`.

mainegra commented:
# 2011 CVS commit for inclusion of EADL atomic relaxations

```
----------------------------
revision 1.74
date: 2011/09/19 22:58:51;  author: mainegra;  state: Exp;  lines: +6 -5
```

Implemented a more accurate simulation of the atomic relaxation process.
Data taken from the Evaluated Atomic Data Library (EADL) from LLNL.
The EADL data is available for EGSnrc in $HEN_HOUSE/data/relax.data
and was created by I. Kawrakow for use with C++ user-codes.

Reference:
D.E. Cullen, et al.,
"Tables and Graphs of Atomic Subshell and Relaxation Data Derived from the
LLNL Evaluated Atomic Data Library (EADL), Z = 1 - 100,"
Lawrence Livermore National Laboratory, UCRL-50400, Vol. 30, October 1991

new subroutines:

subroutine egs_relax_init;
subroutine egs_eadl_relax(iZ, shell_egs);

Sampling the shell for the transition is accomplished using an alias
sampling algorithm. Hence no significant increase of CPU time was
observed.

### Inconsistency in the binding energies:

- COMPT uses variable be_array from file incoh.data
  (Lederer and Shirley 1978). PHOTO, MOLLER (EII) and PHOTO use variable
  binding_energies read from file photo_relax.data. be_array and binding_energies
  are almost identical except binding_energies is limited to average <M> and <N>
  as explained in PIRS-701.

- egs_relax_init reads EADL binding energies in variable shell_be, which seems
  to be somewhat different from the above values.
  (about 0.5% different for W for instance)

Since I implemented this relatively quickly before McGill 2011, I haven't
had time to check whether this inconsistency has any signifficant effect.
For this reason, the old relaxation implementation is left as the default
until further testing/thinking is done. The way to turn on EADL relaxation
is by setting macro $EADL_RELAX to .true. in egsnrc.macros or in the user-code.
It is currently set to .false.
